Optimize KV cache distribution for asymmetric pipeline parallelism

gholmes829 opened the pull request:
<!-- markdownlint-disable -->

## Purpose

I've discovered some troublesome inefficiencies in how KV cache memory is allocated for asymmetric pipeline parallelism.

I use pipeline parallelism with 2 GPUs and asymmetrically load layers between them. When using `--kv-cache-memory-bytes`, VLLM correctly allocates only as much VRAM as is needed to load the model layers on the devices.

...HOWEVER for the KV cache, tensor sizes weren't being adjusted to match the resolved block count. With a 56/8 layer split for example, the second worker calculates 327,680 token capacity (via memory budget), but coordination locks its usable pool to 46,800 tokens (same as first worker) yet still allocates tensors as if for the full 327,680 tokens, thus wasting ~4.3 GiB.

This problem grows directly to the degree of asymmetry and causes system throughput to become bottlenecked by the most constrained device.

**Solution**:

We can address this by tempering the size of the kv cache tensors, namely by shrinking them proportional to the minimum block count across ranks.

This ensures memory allocations actually match the resolved minimum block count (prior to getting allocated).

This fix also requires extracting and moving kv cache config reporting to ensure the resolved values are logged rather than the original pre-shrunk estimations.

## Test Plan

Manually tested with asymmetric PP:
VLLM_PP_LAYER_PARTITION=56,8 vllm serve Qwen/Qwen3-32B \
      --pipeline-parallel-size 2 \
      --kv-cache-memory-bytes 5G

Environment details:
* Docker WSL2
* Online serving VLLM V1
* From `vllm/vllm-openai:v0.10.2`
* Cuda 2x3090

I updated relevant unit tests to correctly model adjusted calculations (please see changes to `tests/v1/core/test_kv_cache_utils.py`).

## Test Result

(A) Baseline (pre-implementation):
```
vllm  | (Worker_PP1 pid=167) INFO 10-06 22:46:40 [gpu_worker.py:267] Initial free memory 22.63 GiB, reserved 5.00 GiB memory for KV Cache as specified by kv_cache_memory_bytes config and skipped memory profiling. This does not respect the gpu_memory_utilization config. Only use kv_cache_memory_bytes config when you want manual control of KV cache memory size. If OOM'ed, check the difference of initial free memory between the current run and the previous run where kv_cache_memory_bytes is suggested and update it correspondingly.
vllm  | (Worker_PP0 pid=166) INFO 10-06 22:48:30 [gpu_worker.py:267] Initial free memory 22.63 GiB, reserved 5.00 GiB memory for KV Cache as specified by kv_cache_memory_bytes config and skipped memory profiling. This does not respect the gpu_memory_utilization config. Only use kv_cache_memory_bytes config when you want manual control of KV cache memory size. If OOM'ed, check the difference of initial free memory between the current run and the previous run where kv_cache_memory_bytes is suggested and update it correspondingly.
vllm  | (EngineCore_DP0 pid=116) INFO 10-06 22:48:31 [kv_cache_utils.py:1267] GPU KV cache size: 46,800 tokens
vllm  | (EngineCore_DP0 pid=116) INFO 10-06 22:48:31 [kv_cache_utils.py:1272] Maximum concurrency for 40,960 tokens per request: 1.14x
vllm  | (EngineCore_DP0 pid=116) INFO 10-06 22:48:31 [kv_cache_utils.py:1267] GPU KV cache size: 327,680 tokens
vllm  | (EngineCore_DP0 pid=116) INFO 10-06 22:48:31 [kv_cache_utils.py:1272] Maximum concurrency for 40,960 tokens per request: 8.00x
```

(B) Fixed PR version:
```
vllm  | (Worker_PP1 pid=167) INFO 10-06 22:25:09 [gpu_worker.py:267] Initial free memory 22.63 GiB, reserved 5.00 GiB memory for KV Cache as specified by kv_cache_memory_bytes config and skipped memory profiling. This does not respect the gpu_memory_utilization config. Only use kv_cache_memory_bytes config when you want manual control of KV cache memory size. If OOM'ed, check the difference of initial free memory between the current run and the previous run where kv_cache_memory_bytes is suggested and update it correspondingly.
vllm  | (Worker_PP0 pid=166) INFO 10-06 22:26:53 [gpu_worker.py:267] Initial free memory 22.63 GiB, reserved 5.00 GiB memory for KV Cache as specified by kv_cache_memory_bytes config and skipped memory profiling. This does not respect the gpu_memory_utilization config. Only use kv_cache_memory_bytes config when you want manual control of KV cache memory size. If OOM'ed, check the difference of initial free memory between the current run and the previous run where kv_cache_memory_bytes is suggested and update it correspondingly.
vllm  | (EngineCore_DP0 pid=116) INFO 10-06 22:26:53 [kv_cache_utils.py:1267] GPU KV cache size: 46,800 tokens
vllm  | (EngineCore_DP0 pid=116) INFO 10-06 22:26:53 [kv_cache_utils.py:1272] Maximum concurrency for 40,960 tokens per request: 1.14x
vllm  | (EngineCore_DP0 pid=116) INFO 10-06 22:26:53 [kv_cache_utils.py:1267] GPU KV cache size: 46,800 tokens
vllm  | (EngineCore_DP0 pid=116) INFO 10-06 22:26:53 [kv_cache_utils.py:1272] Maximum concurrency for 40,960 tokens per request: 1.14x
```

Note how for Scenario A it allocates:
* GPU0 -- 46,800 tokens @1.14x concurrency -> allocates ~5GB (all available)
* GPU1 -- 327,680 tokens @8.00x concurrency -> allocates ~5GB (~0.7GB usable with ~4.3GB inaccessible)

And for Scenario B it allocates:
* GPU0 -- 46,800 tokens @1.14x concurrency  -> allocates ~5GB (all available)
* GPU1 -- 46,800 tokens @1.14x concurrency  -> allocates ~0.7GB (all available)

In terms of numbers, in my scenario this **reduces KV cache memory usage by ~40%** for the same performance! Other setups could benefit more or less than this depending on the partitioning and concurrency multiplier.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [X] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [X] The test plan, such as providing test command.
- [X] The test results, such as pasting the results comparison before and after, or e2e results
- [X] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

gholmes829 committed:
Add proportional KV cache distribution for asymmetric pipeline parallelism

Signed-off-by: gholmes829 <g.holmes429@gmail.com>

gholmes829 committed:
Simplify layer count calculation

Signed-off-by: gholmes829 <g.holmes429@gmail.com>

gholmes829 committed:
Fix line length violations

Signed-off-by: gholmes829 <g.holmes429@gmail.com>

gholmes829 commented:
Rebased on main

youkaichao commented:
cc @heheda12345

heheda12345 reviewed (commented):
Nice catch! But I think we don't need to add a flag to control this.

https://github.com/vllm-project/vllm/blob/59f30d0448999eca6b2428e6985491412596f261/vllm/v1/core/kv_cache_utils.py#L1351-L1355

See the code here. The available num_blocks is the minimum among all PP rank. However, we only reduce the num_block for each rank, but didn't reduce the kv_cache_memory size `KVCacheTensor.size` in https://github.com/vllm-project/vllm/blob/59f30d0448999eca6b2428e6985491412596f261/vllm/v1/kv_cache_interface.py#L389

So the rank with less amount of memory allocates some useless memory.

I think you can fix the problem by something like
```
num_blocks_old = kv_cache_config.num_blocks
for tensor in kv_cache_config.kv_cache_tensors:
     assert tensor.size % num_blocks_old == 0
     tensor.size = tensor.size // num_blocks_old * num_blocks
```

Could you have a try?

gholmes829 committed:
Adjust kv cache proportionally by shrinking tensor size

Signed-off-by: gholmes829 <g.holmes429@gmail.com>

heheda12345 reviewed (approved):
LGTM! Thanks for the fix.

heheda12345 commented:
@gholmes829 can you update the PR description for the new approach?

gholmes829 committed:
Fix kv cache size and concurrency reporting

Signed-off-by: gholmes829 <g.holmes429@gmail.com>

gholmes829 committed:
Fix incorrect test values

Signed-off-by: gholmes829 <g.holmes429@gmail.com>

gholmes829 committed:
Formatting

Signed-off-by: gholmes829 <g.holmes429@gmail.com>

gholmes829 commented:
@heheda12345 I've pushed rest of changes and updated PR description.

Code changes since your review:
* Moved kv cache config logging later so it reflects tensor size adjustment
* Updated expected values of tests
* Formatting

Everything looks good and ready from my end, barring ci completion. Let me know if anything else needed!
